week-4-tests-and-persistence/w04e02-test-nest/src/users/auth/auth.service.spec.ts

```diff
@@ -1,12 +1,14 @@
 import { Test, TestingModule } from '@nestjs/testing';
+import { PrismaService } from '../../database/prisma.service';
+import { UsersService } from '../users.service';
 import { AuthService } from './auth.service';
 
 describe('AuthService', () => {
   let service: AuthService;
 
   beforeEach(async () => {
     const module: TestingModule = await Test.createTestingModule({
-      providers: [AuthService],
+      providers: [AuthService, UsersService, PrismaService],
```

Comment:
Tutaj nie może być faktyczna implementacja `PrismaService` ponieważ wtedy pracujesz na faktycznej bazie danych.

W tym układzie testy zrobią się bardzo "ciężkie" ponieważ nie będą to już `unit` a `integration` tests.

Tego nie chcemy.

Musisz dostarczyć Stuba już na poziomie `UsersService` zobacz:
```typescript
{
          provide: UsersService,
          useValue: {}
}
```

Zauważ że o ile test nr. 1:
_'should throw error when wrong credentials'_

Zadziała poprawnie, to ten `2` w tym układzie (jak mam stub) - się wykolei, `3` zresztą też.

Dzieje się tak dlatego - ponieważ musimy do tego `useValue` dopisać 2 metody jeszcze - żeby wszystko działało.